### PR TITLE
Fix #2830: Add Night Mode

### LINKED ISSUE #2830: Add Night Mode
Hi All,

Is there any plan to add Night Mode on Brave iOS?

The example of Night Mode is as in Firefox iOS below, Brave iOS is based on Firefox iOS, thus it should be easier to fork Night Mode from Firefox iOS and add to Brave iOS. Night Mode is different than Dark Mode or Reader Mode.

Turn on Night Mode:

WhatsApp Image 2020-08-21 at 09 22 30

From this:

WhatsApp Image 2020-08-21 at 09 22 30 (1)

Into this:

WhatsApp Image 2020-08-21 at 09 22 30 (2)

Thank you.

### PR DESCRIPTION
**Pull request by soner-yuksel**
## Summary of Changes

This PR is adding Night Mode injection to our browser. The script is from Firefox [NightModeHelper](https://github.com/mozilla-mobile/firefox-ios/blob/main/Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/NightModeHelper.js)
and Night Mode is integrated accordingly with Dark/Light Mode.

This pull request fixes #2830

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:

- Go Settings -> Appearance -> Set APPEARANCE light Mode
- Browse wikipedia.org
- Check Main Page on website is not in night mode and application itself is light mode on
- Go back to Appearance Settings and Set Night mode toggle on 
- Check Application itself goes to dark mode and check open website is in night mode
- Go back to Appearance Settings and Set Night mode toggle off 
- Check Application itself goes to light mode and check open website is not in night mode

So the idea here is When NightMode is enabled it will also effect Application Appearance and it will change application to dark mode and when the Night mode toggle is off, it will switch back to the previous preference. 

## Screenshots:

test

https://user-images.githubusercontent.com/6643505/157545057-ca0addb9-e549-4a61-967b-9976a2352e79.MP4

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

**Review by Brandon-T — APPROVED** [review]

**Comment by iccub** [comment]
Blocking for now, i have design concerns over this

File: BraveShared/BraveStrings.swift [hunk 1]
```diff
@@ -3593,6 +3593,37 @@ extension Strings {
...
 }
 
+// MARK: - Night Mode
+
+extension Strings {
+    public struct NightMode {
+        public static let sectionTitle = NSLocalizedString(
+            "nightMode.modeTitle",
+            bundle: .braveShared,
+            value: "MODE",
```
Comment (iccub) [inline BraveShared/BraveStrings.swift]: Write it the normal way, then we can uppercase it in code

File: Client/Application/ClientPreferences.swift [hunk 2]
```diff
@@ -68,6 +68,8 @@ extension Preferences {
         /// Defines the user's normal browsing theme
         /// `system`, follows the current OS display mode
         static let themeNormalMode = Option<String>(key: "general.normal-mode-theme", default: DefaultTheme.system.rawValue)
+        /// Specifies If Night Mode is enabled
+        static let nightModeEnabled = Option<Bool>(key: "general.night-mode-theme", default: false)
```
Comment (iccub) [inline Client/Application/ClientPreferences.swift]: nitpick, key should be `general.night-mode-enabled`

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/MenuViewController.swift [hunk 3]
```diff
@@ -12,22 +12,24 @@ import SwiftUI
 
 struct MenuItemHeaderView: View {
     @Environment(\.colorScheme) private var colorScheme: ColorScheme
-    var icon: UIImage
+    var icon: UIImage?
```
Comment (iccub) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/MenuViewController.swift]: I don't think this has to be optional? in code below you fallback to an empty UIImage regardless

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/NightModeMenuButton.swift [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        .frame(maxWidth: .infinity, minHeight: 48.0, alignment: .leading)
+        .background(
+            Button(action: { toggleNightMode(!Preferences.General.nightModeEnabled.value) }) {
+                Color.clear
+            }
+            .buttonStyle(TableCellButtonStyle())
+        )
+        .accessibilityElement()
+        .accessibility(addTraits: .isButton)
+        .accessibility(label: Text("Night Mode"))
```
Comment (iccub) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/NightModeMenuButton.swift]: we can use localized string for that, i think you simply copied from the vpn cell, for that case `Brave VPN` is not supposed to be translated at all

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/NightModeMenuButton.swift [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        )
+    }
+    
+    private func toggleNightMode(_ enabled: Bool) {
+        changeNightModePreference(enabled)
+    }
+    
+    private var actionToggle: some View {
+        Group {
+            let toggle = Toggle("Action Toggle", isOn: isNightModeEnabledBinding)
```
Comment (iccub) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/NightModeMenuButton.swift]: This text should not be needed

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/NightModeMenuButton.swift [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            set: { toggleNightMode($0) }
+        )
+    }
+    
+    private func toggleNightMode(_ enabled: Bool) {
+        changeNightModePreference(enabled)
+    }
+    
+    private var actionToggle: some View {
+        Group {
```
Comment (iccub) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/NightModeMenuButton.swift]: I think you could remove the `Group` here and the separate declaration, place it like

```
Toggle("Action Toggle", isOn: isNightModeEnabledBinding)
.labelsHidden()
.toggleStyle(SwitchToggleStyle(tint: .accentColor))
```

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/NightModeMenuButton.swift [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        HStack {
+            MenuItemHeaderView(
+                icon: UIImage(systemName: "moon")?.template ?? UIImage(),
+                title: Strings.NightMode.settingsTitle,
+                subtitle: Strings.NightMode.settingsDescription)
+            Spacer()
+            actionToggle
+        }
+        .padding(.horizontal, 14)
+        .frame(maxWidth: .infinity, minHeight: 48.0, alignment: .leading)
```
Comment (iccub) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/NightModeMenuButton.swift]: `alignment` needed here?

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/NightModeMenuButton.swift [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    
+    private var isNightModeEnabledBinding: Binding<Bool> {
+        Binding(
+            get: { isNightModeEnabled },
+            set: { toggleNightMode($0) }
+        )
+    }
+    
+    private func toggleNightMode(_ enabled: Bool) {
+        changeNightModePreference(enabled)
```
Comment (iccub) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/NightModeMenuButton.swift]: I think all of this mode changing logic could be removed,
instead we could add a preference observer directly to the TabManager,
and if night mode is detected we call the `NightModeHelper.setNightMode`

**Review by iccub — CHANGES_REQUESTED** [review]

**Review by iccub — APPROVED** [review]

**Comment by TaxingAuthority** [comment]
Is it possible to have the option for the new night mode to turn on and off with the iOS setting? As it is now the toggle stays in night mode when iOS  changes to light appearance. Thanks for all the work!

**Pull request merged**